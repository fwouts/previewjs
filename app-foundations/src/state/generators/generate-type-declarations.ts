import type { CollectedTypes, ValueType } from "@previewjs/type-analyzer";
import assertNever from "assert-never";

export function generateTypeDeclarations(
  name: string,
  type: ValueType,
  argKeys: string[],
  collected: CollectedTypes
) {
  const typeNameMapping: {
    [name: string]: string;
  } = {};
  const usedTypes = new Set<string>();
  const declaredTypes = new Set<string>();
  let hasProperType: boolean;
  let propsTypeName: string;
  if (type.kind === "name") {
    hasProperType = true;
    propsTypeName = type.name;
    usedTypes.add(type.name);
  } else {
    hasProperType = false;
    propsTypeName = `:${name}Props`;
  }
  const safePropsTypeName = safeTypeName(propsTypeName, typeNameMapping);
  let output = `declare let properties: { children?: any } & ${
    argKeys.length === 0
      ? safePropsTypeName
      : `__previewjs_Optionalize<${safePropsTypeName}, ${argKeys
          .map((prop) => `"${prop.replace(/"/g, '\\"')}"`)
          .join(" | ")}>`
  };

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;`;

  if (!hasProperType) {
    output += `

type ${safePropsTypeName} = ${generateTypeScriptType(
      type,
      collected,
      typeNameMapping,
      usedTypes,
      true
    )};`;
  }

  let generatedCount: number;
  do {
    generatedCount = 0;
    for (const typeName of [...usedTypes]) {
      if (declaredTypes.has(typeName)) {
        continue;
      }
      const type = collected[typeName];
      if (!type) {
        continue;
      }
      let params = "";
      if (Object.keys(type.parameters).length > 0) {
        params = "<";
        for (const [paramName, defaultType] of Object.entries(
          type.parameters
        )) {
          params += paramName;
          if (defaultType) {
            params += ` = ${generateTypeScriptType(
              defaultType,
              collected,
              typeNameMapping,
              usedTypes
            )}`;
          }
          params += ",";
        }
        params += ">";
      }
      output += `

type ${safeTypeName(
        typeName,
        typeNameMapping
      )}${params} = ${generateTypeScriptType(
        type.type,
        collected,
        typeNameMapping,
        usedTypes,
        typeName === propsTypeName
      )}`;
      declaredTypes.add(typeName);
      generatedCount += 1;
    }
  } while (generatedCount > 0);

  return output;
}

function generateTypeScriptType(
  type: ValueType,
  collected: CollectedTypes,
  typeNameMapping: {
    [name: string]: string;
  },
  usedTypes: Set<string>,
  optionalFunctionProps = false
): string {
  switch (type.kind) {
    case "any":
      return "any";
    case "unknown":
      return "unknown";
    case "never":
      return "never";
    case "void":
      return "void";
    case "null":
      return "null";
    case "boolean":
      return "boolean";
    case "string":
      return "string";
    case "number":
      return "number";
    case "node":
      // Since React type definitions aren't included in the editor, we need to use "any" here.
      return "any";
    case "literal":
      if (typeof type.value === "number") {
        return type.value.toString(10);
      } else if (typeof type.value === "string") {
        return `"${type.value.replace(/"/g, '\\"')}"`;
      } else {
        return type.value ? "true" : "false";
      }
    case "enum":
      return Object.values(type.options)
        .map((value) =>
          typeof value === "number"
            ? value.toString(10)
            : `"${value.replace(/"/g, '\\"')}"`
        )
        .join(" | ");
    case "array":
      return `Array<${generateTypeScriptType(
        type.items,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "set":
      return `Set<${generateTypeScriptType(
        type.items,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "object":
      return `{
              ${Object.entries(type.fields)
                .filter(([propName]) => propName !== "children")
                .map(([propName, propType]) => {
                  const resolvedPropType =
                    propType.kind === "name"
                      ? collected[propType.name]?.type || propType
                      : propType;
                  const optional =
                    (optionalFunctionProps &&
                      resolvedPropType?.kind === "function") ||
                    resolvedPropType.kind === "optional" ||
                    resolvedPropType.kind === "any" ||
                    resolvedPropType.kind === "unknown";
                  return `["${propName.replace(/"/g, '\\"')}"]${
                    optional ? "?" : ""
                  }: ${generateTypeScriptType(
                    propType,
                    collected,
                    typeNameMapping,
                    usedTypes
                  )}`;
                })
                .join("\n")}
            }`;
    case "map":
      return `Map<${generateTypeScriptType(
        type.keys,
        collected,
        typeNameMapping,
        usedTypes
      )}, ${generateTypeScriptType(
        type.values,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "record":
      return `Record<${generateTypeScriptType(
        type.keys,
        collected,
        typeNameMapping,
        usedTypes
      )}, ${generateTypeScriptType(
        type.values,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "union":
      return type.types
        .map(
          (subtype) =>
            `(${generateTypeScriptType(
              subtype,
              collected,
              typeNameMapping,
              usedTypes
            )})`
        )
        .join(" | ");
    case "intersection":
      return type.types
        .map(
          (subtype) =>
            `(${generateTypeScriptType(
              subtype,
              collected,
              typeNameMapping,
              usedTypes
            )})`
        )
        .join(" & ");
    case "function":
      return `(...params: any[]) => (${generateTypeScriptType(
        type.returnType,
        collected,
        typeNameMapping,
        usedTypes
      )})`;
    case "optional":
      return `${generateTypeScriptType(
        type.type,
        collected,
        typeNameMapping,
        usedTypes
      )} | undefined`;
    case "promise":
      return `Promise<${generateTypeScriptType(
        type.type,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "name":
      usedTypes.add(type.name);
      let name = safeTypeName(type.name, typeNameMapping);
      if (type.args.length > 0) {
        name += "<";
        for (let i = 0; i < type.args.length; i++) {
          if (i > 0) {
            name += ",";
          }
          const arg = type.args[i]!;
          name += generateTypeScriptType(
            arg,
            collected,
            typeNameMapping,
            usedTypes
          );
        }
        name += ">";
      }
      return name;
    default:
      throw assertNever(type);
  }
}

function safeTypeName(
  fullTypeName: string,
  typeNameMapping: {
    [name: string]: string;
  }
) {
  const existingName = typeNameMapping[fullTypeName];
  if (existingName) {
    return existingName;
  }
  const typeName = fullTypeName.split(":")[1]!;
  if (!typeName) {
    // This is a generic type name.
    return fullTypeName;
  }
  let i = 1;
  const existingTypeNames = new Set(Object.values(typeNameMapping));
  while (existingTypeNames.has(numberedName(typeName, i))) {
    i++;
  }
  const result = numberedName(typeName, i);
  typeNameMapping[fullTypeName] = result;
  return result;
}

function numberedName(typeName: string, i: number) {
  return `${typeName}${i === 1 ? "" : `_${i}`}`;
}
