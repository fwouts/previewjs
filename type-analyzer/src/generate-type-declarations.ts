import assertNever from "assert-never";
import prettier from "prettier";
import parserTypescript from "prettier/parser-typescript";
import type { CollectedTypes, ValueType } from "./definitions";

export function generateTypeDeclarations(
  names: string[],
  collected: CollectedTypes
) {
  const typeNameMapping: {
    [name: string]: string;
  } = {};
  const usedTypes = new Set(names);
  const declaredTypes = new Set<string>();

  let output = "";
  let generatedCount: number;
  do {
    generatedCount = 0;
    for (const typeName of [...usedTypes]) {
      if (declaredTypes.has(typeName)) {
        continue;
      }
      const type = collected[typeName];
      if (!type) {
        continue;
      }
      let params = "";
      if (Object.keys(type.parameters).length > 0) {
        params = "<";
        for (const [paramName, defaultType] of Object.entries(
          type.parameters
        )) {
          params += paramName;
          if (defaultType) {
            params += ` = ${generateTypeScriptType(
              defaultType,
              collected,
              typeNameMapping,
              usedTypes
            )}`;
          }
          params += ",";
        }
        params += ">";
      }
      output += `

type ${safeTypeName(
        typeName,
        typeNameMapping
      )}${params} = ${generateTypeScriptType(
        type.type,
        collected,
        typeNameMapping,
        usedTypes
      )}`;
      declaredTypes.add(typeName);
      generatedCount += 1;
    }
  } while (generatedCount > 0);

  return prettier
    .format(output, {
      parser: "typescript",
      plugins: [parserTypescript],
      filepath: "types.ts",
      trailingComma: "none",
    })
    .trim();
}

function generateTypeScriptType(
  type: ValueType,
  collected: CollectedTypes,
  typeNameMapping: {
    [name: string]: string;
  },
  usedTypes: Set<string>
): string {
  switch (type.kind) {
    case "any":
      return "any";
    case "unknown":
      return "unknown";
    case "never":
      return "never";
    case "void":
      return "void";
    case "null":
      return "null";
    case "boolean":
      return "boolean";
    case "string":
      return "string";
    case "number":
      return "number";
    case "node":
      // Since React type definitions aren't included in the editor, we need to use "any" here.
      return "any";
    case "literal":
      if (typeof type.value === "number") {
        return type.value.toString(10);
      } else if (typeof type.value === "string") {
        return `"${type.value.replace(/"/g, '\\"')}"`;
      } else {
        return type.value ? "true" : "false";
      }
    case "enum":
      return Object.values(type.options)
        .map((value) =>
          typeof value === "number"
            ? value.toString(10)
            : `"${value.replace(/"/g, '\\"')}"`
        )
        .join(" | ");
    case "array":
      return `Array<${generateTypeScriptType(
        type.items,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "set":
      return `Set<${generateTypeScriptType(
        type.items,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "tuple":
      return `[${type.items
        .map((item) =>
          generateTypeScriptType(item, collected, typeNameMapping, usedTypes)
        )
        .join(", ")}]`;
    case "object":
      return `{
              ${Object.entries(type.fields)
                .map(([propName, propType]) => {
                  const resolvedPropType =
                    propType.kind === "name"
                      ? collected[propType.name]?.type || propType
                      : propType;
                  const optional =
                    resolvedPropType.kind === "optional" ||
                    resolvedPropType.kind === "any" ||
                    resolvedPropType.kind === "unknown";
                  return `["${propName.replace(/"/g, '\\"')}"]${
                    optional ? "?" : ""
                  }: ${generateTypeScriptType(
                    propType,
                    collected,
                    typeNameMapping,
                    usedTypes
                  )}`;
                })
                .join("\n")}
            }`;
    case "map":
      return `Map<${generateTypeScriptType(
        type.keys,
        collected,
        typeNameMapping,
        usedTypes
      )}, ${generateTypeScriptType(
        type.values,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "record":
      return `Record<${generateTypeScriptType(
        type.keys,
        collected,
        typeNameMapping,
        usedTypes
      )}, ${generateTypeScriptType(
        type.values,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "union":
      return type.types
        .map(
          (subtype) =>
            `(${generateTypeScriptType(
              subtype,
              collected,
              typeNameMapping,
              usedTypes
            )})`
        )
        .join(" | ");
    case "intersection":
      return type.types
        .map(
          (subtype) =>
            `(${generateTypeScriptType(
              subtype,
              collected,
              typeNameMapping,
              usedTypes
            )})`
        )
        .join(" & ");
    case "function":
      return `(...params: any[]) => (${generateTypeScriptType(
        type.returnType,
        collected,
        typeNameMapping,
        usedTypes
      )})`;
    case "optional":
      return `${generateTypeScriptType(
        type.type,
        collected,
        typeNameMapping,
        usedTypes
      )} | undefined`;
    case "promise":
      return `Promise<${generateTypeScriptType(
        type.type,
        collected,
        typeNameMapping,
        usedTypes
      )}>`;
    case "name": {
      usedTypes.add(type.name);
      let name = safeTypeName(type.name, typeNameMapping);
      if (type.args.length > 0) {
        name += "<";
        for (let i = 0; i < type.args.length; i++) {
          if (i > 0) {
            name += ",";
          }
          const arg = type.args[i]!;
          name += generateTypeScriptType(
            arg,
            collected,
            typeNameMapping,
            usedTypes
          );
        }
        name += ">";
      }
      return name;
    }
    default:
      throw assertNever(type);
  }
}

function safeTypeName(
  fullTypeName: string,
  typeNameMapping: {
    [name: string]: string;
  }
) {
  const existingName = typeNameMapping[fullTypeName];
  if (existingName) {
    return existingName;
  }
  const typeName = fullTypeName.split(":")[1]!;
  if (!typeName) {
    // This is a generic type name.
    return fullTypeName;
  }
  let i = 1;
  const existingTypeNames = new Set(Object.values(typeNameMapping));
  while (existingTypeNames.has(numberedName(typeName, i))) {
    i++;
  }
  const result = numberedName(typeName, i);
  typeNameMapping[fullTypeName] = result;
  return result;
}

function numberedName(typeName: string, i: number) {
  return `${typeName}${i === 1 ? "" : `_${i}`}`;
}
