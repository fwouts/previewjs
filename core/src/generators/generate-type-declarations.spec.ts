import {
  ANY_TYPE,
  arrayType,
  BOOLEAN_TYPE,
  enumType,
  functionType,
  intersectionType,
  literalType,
  namedType,
  NEVER_TYPE,
  NODE_TYPE,
  NULL_TYPE,
  NUMBER_TYPE,
  objectType,
  optionalType,
  promiseType,
  recordType,
  setType,
  STRING_TYPE,
  unionType,
  UNKNOWN_TYPE,
  VOID_TYPE,
} from "@previewjs/type-analyzer";
import { generateTypeDeclarations } from "./generate-type-declarations";

describe("generateTypeDeclarations", () => {
  test("simple props with object type", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        objectType({
          foo: STRING_TYPE,
        }),

        new Set([]),
        {}
      )
    ).toMatchSnapshot();
  });

  test("simple props with named type", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:MyComponentProps"),
        new Set([]),
        {
          "/foo.tsx:MyComponentProps": {
            type: objectType({
              foo: STRING_TYPE,
            }),
            parameters: {},
          },
        }
      )
    ).toMatchSnapshot();

    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:Foo"),
        new Set([]),
        {
          "/foo.tsx:Foo": {
            type: objectType({
              foo: STRING_TYPE,
            }),
            parameters: {},
          },
        }
      )
    ).toMatchSnapshot();
  });

  test("recursive named type", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:Foo"),
        new Set([]),
        {
          "/foo.tsx:Foo": {
            type: objectType({
              child: optionalType(namedType("/foo.tsx:Foo")),
            }),
            parameters: {},
          },
        }
      )
    ).toMatchSnapshot();
  });

  test("recursive function type", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:MyComponentProps"),
        new Set([]),
        {
          "/foo.tsx:MyComponentProps": {
            type: objectType({
              foo: namedType("/foo.tsx:Fn"),
            }),
            parameters: {},
          },
          "/foo.tsx:Fn": {
            type: functionType(optionalType(namedType("/foo.tsx:Fn"))),
            parameters: {},
          },
        }
      )
    ).toMatchSnapshot();
  });

  test("all types", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:Foo"),
        new Set([]),
        {
          "/foo.tsx:Foo": {
            type: objectType({
              anyType: ANY_TYPE,
              unknownType: UNKNOWN_TYPE,
              neverType: NEVER_TYPE,
              voidType: VOID_TYPE,
              nullType: NULL_TYPE,
              booleanType: BOOLEAN_TYPE,
              stringType: STRING_TYPE,
              numberType: NUMBER_TYPE,
              reactNodeType: NODE_TYPE,
              numberLiteral: literalType(123),
              stringLiteral: literalType("foo"),
              trueLiteral: literalType(true),
              falseLiteral: literalType(false),
              stringEnumType: enumType({
                a: "A",
                b: "B",
                c: "C",
              }),
              numberEnumType: enumType({
                a: 3,
                b: 2,
                c: 1,
              }),
              arrayType: arrayType(STRING_TYPE),
              setType: setType(STRING_TYPE),
              recordType: recordType(STRING_TYPE, NUMBER_TYPE),
              unionType: unionType([STRING_TYPE, NUMBER_TYPE]),
              intersectionType: intersectionType([STRING_TYPE, NUMBER_TYPE]),
              functionType: functionType(STRING_TYPE),
              promiseType: promiseType(STRING_TYPE),
              namedType: namedType("/foo.tsx:Bar"),
            }),
            parameters: {},
          },
          "/foo.tsx:Bar": {
            type: objectType({
              bar: functionType(STRING_TYPE),
            }),
            parameters: {},
          },
        }
      )
    ).toMatchSnapshot();
  });

  test("duplicate types in different files", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:Foo"),
        new Set([]),
        {
          "/foo.tsx:Foo": {
            type: objectType({
              foo: namedType("/bar.tsx:Foo"),
            }),
            parameters: {},
          },
          "/bar.tsx:Foo": {
            type: objectType({
              bar: namedType("/baz.tsx:Foo"),
            }),
            parameters: {},
          },
          "/baz.tsx:Foo": {
            type: objectType({
              baz: STRING_TYPE,
            }),
            parameters: {},
          },
        }
      )
    ).toMatchSnapshot();
  });

  test("optional props in object type", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        objectType({
          a: STRING_TYPE,
          b: optionalType(STRING_TYPE),
          c: ANY_TYPE,
          d: UNKNOWN_TYPE,
          e: STRING_TYPE,
          f: optionalType(STRING_TYPE),
        }),
        new Set([]),
        {}
      )
    ).toMatchSnapshot();
  });

  test("generic types", () => {
    expect(
      generateTypeDeclarations(
        "MyComponent",
        namedType("/foo.tsx:A"),
        new Set([]),
        {
          "/foo.tsx:A": {
            type: namedType("T"),
            parameters: { T: namedType("/foo.tsx:B") },
          },
          "/foo.tsx:B": {
            type: objectType({
              foo: namedType("/foo.tsx:C", [namedType("/foo.tsx:B")]),
            }),
            parameters: {},
          },
          "/foo.tsx:C": {
            type: objectType({
              t: namedType("T"),
              s: namedType("S"),
            }),
            parameters: {
              T: null,
              S: namedType("T"),
            },
          },
        }
      )
    ).toMatchSnapshot();
  });
});
