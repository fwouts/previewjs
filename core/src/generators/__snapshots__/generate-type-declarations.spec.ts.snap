// Vitest Snapshot v1

exports[`generateTypeDeclarations > all types 1`] = `
"declare let properties: { children?: any } & Foo;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type Foo = {
              [\\"anyType\\"]?: any
[\\"unknownType\\"]?: unknown
[\\"neverType\\"]: never
[\\"voidType\\"]: void
[\\"nullType\\"]: null
[\\"booleanType\\"]: boolean
[\\"stringType\\"]: string
[\\"numberType\\"]: number
[\\"reactNodeType\\"]: any
[\\"numberLiteral\\"]: 123
[\\"stringLiteral\\"]: \\"foo\\"
[\\"trueLiteral\\"]: true
[\\"falseLiteral\\"]: false
[\\"stringEnumType\\"]: \\"A\\" | \\"B\\" | \\"C\\"
[\\"numberEnumType\\"]: 3 | 2 | 1
[\\"arrayType\\"]: Array<string>
[\\"setType\\"]: Set<string>
[\\"recordType\\"]: Record<string, number>
[\\"unionType\\"]: (string) | (number)
[\\"intersectionType\\"]: (string) & (number)
[\\"functionType\\"]?: (...params: any[]) => (string)
[\\"promiseType\\"]: Promise<string>
[\\"namedType\\"]: Bar
            }

type Bar = {
              [\\"bar\\"]: (...params: any[]) => (string)
            }"
`;

exports[`generateTypeDeclarations > duplicate types in different files 1`] = `
"declare let properties: { children?: any } & Foo;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type Foo = {
              [\\"foo\\"]: Foo_2
            }

type Foo_2 = {
              [\\"bar\\"]: Foo_3
            }

type Foo_3 = {
              [\\"baz\\"]: string
            }"
`;

exports[`generateTypeDeclarations > generic types 1`] = `
"declare let properties: { children?: any } & A;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type A<T = B,> = T

type B = {
              [\\"foo\\"]: C<B>
            }

type C<T,S = T,> = {
              [\\"t\\"]: T
[\\"s\\"]: S
            }"
`;

exports[`generateTypeDeclarations > optional props in object type 1`] = `
"declare let properties: { children?: any } & MyComponentProps;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type MyComponentProps = {
              [\\"a\\"]: string
[\\"b\\"]?: string | undefined
[\\"c\\"]?: any
[\\"d\\"]?: unknown
[\\"e\\"]: string
[\\"f\\"]?: string | undefined
            };"
`;

exports[`generateTypeDeclarations > recursive function type 1`] = `
"declare let properties: { children?: any } & MyComponentProps;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type MyComponentProps = {
              [\\"foo\\"]?: Fn
            }

type Fn = (...params: any[]) => (Fn | undefined)"
`;

exports[`generateTypeDeclarations > recursive named type 1`] = `
"declare let properties: { children?: any } & Foo;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type Foo = {
              [\\"child\\"]?: Foo | undefined
            }"
`;

exports[`generateTypeDeclarations > simple props with named type 1`] = `
"declare let properties: { children?: any } & MyComponentProps;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type MyComponentProps = {
              [\\"foo\\"]: string
            }"
`;

exports[`generateTypeDeclarations > simple props with named type 2`] = `
"declare let properties: { children?: any } & Foo;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type Foo = {
              [\\"foo\\"]: string
            }"
`;

exports[`generateTypeDeclarations > simple props with object type 1`] = `
"declare let properties: { children?: any } & MyComponentProps;

type __previewjs_Optionalize<T, K extends keyof T> = {
  [P in Exclude<__previewjs_RequiredKeys<T>, K>]: T[P]
} & {
  [P in Exclude<__previewjs_OptionalKeys<T>, K>]?: T[P]
} & {
  [P in K]?: T[P]
}
type __previewjs_RequiredKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? never : K }[keyof T];
type __previewjs_OptionalKeys<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? K : never }[keyof T];

declare function fn<T>(name: string, returnValue?: T): () => T;

type MyComponentProps = {
              [\\"foo\\"]: string
            };"
`;
